### PR TITLE
Improve LLM model tool annotation

### PR DESCRIPTION
**Pull request by dghirardo**
### Issue #637

This pull request implements the `tool_tailor` gem (https://github.com/kieranklaassen/tool_tailor) to automate the generation of tool annotations.

**Changes**:

- [x] All previously used JSON files for tool annotations have been removed.
- [x] Tool-specific directories have been eliminated, resulting in a cleaner and more straightforward directory structure.
- [x] The `:ANNOTATIONS_PATH` constant has been replaced with `:FUNCTIONS`, an array of symbols corresponding to the tool's functions.
- [x] Methods for validating constants (`:NAME`, `:FUNCTIONS`) have been implemented to ensure consistency and correctness.

### CODE REVIEW

File: lib/langchain/tool/base.rb [hunk 1]
```diff
@@ -97,11 +108,29 @@ def to_google_gemini_tools
     #
     # @return [Hash] Tool's method annotations
     def method_annotations
-      JSON.parse(
-        File.read(
-          self.class.const_get(:ANNOTATIONS_PATH)
-        )
-      )
+      functions.map do |function|
```
Comment (andreibondarev) [inline lib/langchain/tool/base.rb:111]: What do you think about just using all of the `public_instance_methods`?

File: lib/langchain/tool/calculator.rb [hunk 2]
```diff
@@ -13,9 +13,11 @@ class Calculator < Base
     #     calculator = Langchain::Tool::Calculator.new
     #
     NAME = "calculator"
-    ANNOTATIONS_PATH = Langchain.root.join("./langchain/tool/#{NAME}/#{NAME}.json").to_path
```
Comment (andreibondarev) [inline lib/langchain/tool/calculator.rb:16]: 👍🏻

**Comment by andreibondarev** [comment]
Before:
```ruby
irb(main):013> Langchain::Tool::NewsRetriever.new(api_key: ENV["NEWS_API_KEY"]).to_openai_tools
=>
[{"type"=>"function",
  "function"=>
   {"name"=>"news_retriever__get_everything",
    "description"=>"Retrieve all news",
    "parameters"=>
     {"type"=>"object",
      "properties"=>
       {"q"=>{"type"=>"string", "description"=>"Keywords or phrases to search for in the article title and body."},
        "search_in"=>{"type"=>"string", "description"=>"The fields to restrict your q search to. The possible options are: title, description, content."},
        "sources"=>
         {"type"=>"string",
          "description"=>"A comma-seperated string of identifiers (maximum 20) for the news sources or blogs you want headlines from. Use the /sources endpoint to locate these programmatically or look at the sources index."},
        "domains"=>{"type"=>"string", "description"=>"A comma-seperated string of domains (eg bbc.co.uk, techcrunch.com, engadget.com) to restrict the search to."},
        "exclude_domains"=>{"type"=>"string", "description"=>"A comma-seperated string of domains (eg bbc.co.uk, techcrunch.com, engadget.com) to remove from the results."},
        "from"=>{"type"=>"string", "description"=>"A date and optional time for the oldest article allowed. This should be in ISO 8601 format."},
        "to"=>{"type"=>"string", "description"=>"A date and optional time for the newest article allowed. This should be in ISO 8601 format."},
        "language"=>{"type"=>"string", "description"=>"The 2-letter ISO-639-1 code of the language you want to get headlines for. Possible options: ar, de, en, es, fr, he, it, nl, no, pt, ru, se, ud, zh."},
        "sort_by"=>{"type"=>"string", "description"=>"The order to sort the articles in. Possible options: relevancy, popularity, publishedAt."},
        "page_size"=>{"type"=>"integer", "description"=>"The number of results to return per page. 20 is the API's default, 100 is the maximum. Our default is 5."},
        "page"=>{"type"=>"integer", "description"=>"Use this to page through the results."}},
      "required"=>[]}}},
 {"type"=>"function",
  "function"=>
   {"name"=>"news_retriever__get_top_headlines",
    "description"=>"Retrieve top headlines",
    "parameters"=>
     {"type"=>"object",
      "properties"=>
       {"country"=>
         {"type"=>"string",
          "description"=>
           "The 2-letter ISO 3166-1 code of the country you want to get headlines for. Possible options: ae, ar, at, au, be, bg, br, ca, ch, cn, co, cu, cz, de, eg, fr, gb, gr, hk, hu, id, ie, il, in, it, jp, kr, lt, lv, ma, mx, my, ng, nl, no, nz, ph, pl, pt, ro, rs, ru, sa, se, sg, si, sk, th, tr, tw, ua, us, ve, za."},
        "category"=>{"type"=>"string", "description"=>"The category you want to get headlines for. Possible options: business, entertainment, general, health, science, sports, technology."},
        "sources"=>{"type"=>"string", "description"=>"A comma-seperated string of identifiers for the news sources or blogs you want headlines from. Use the /sources endpoint to locate these programmatically."},
        "q"=>{"type"=>"string", "description"=>"Keywords or a phrase to search for."},
        "page_size"=>{"type"=>"integer", "description"=>"The number of results to return per page. 20 is the API's default, 100 is the maximum. Our default is 5."},
        "page"=>{"type"=>"integer", "description"=>"Use this to page through the results."}},
      "required"=>[]}}},
 {"type"=>"function",
  "function"=>
   {"name"=>"news_retriever__get_sources",
    "description"=>"Retrieve news sources",
    "parameters"=>
     {"type"=>"object",
      "properties"=>
       {"category"=>{"type"=>"string", "description"=>"The category you want to get headlines for. Possible options: business, entertainment, general, health, science, sports, technology."},
        "language"=>{"type"=>"string", "description"=>"The 2-letter ISO-639-1 code of the language you want to get headlines for. Possible options: ar, de, en, es, fr, he, it, nl, no, pt, ru, se, ud, zh."},
        "country"=>
         {"type"=>"string",
          "description"=>
           "The 2-letter ISO 3166-1 code of the country you want to get headlines for. Possible options: ae, ar, at, au, be, bg, br, ca, ch, cn, co, cu, cz, de, eg, fr, gb, gr, hk, hu, id, ie, il, in, it, jp, kr, lt, lv, ma, mx, my
, ng, nl, no, nz, ph, pl, pt, ro, rs, ru, sa, se, sg, si, sk, th, tr, tw, ua, us, ve, za."}},
      "required"=>[]}}}]
```

File: lib/langchain/tool/base.rb (hunk 1 above)
Comment (dghirardo) [reply lib/langchain/tool/base.rb:111]: Yes, that could be a good idea, I had already considered it. However, I noticed the `execute_search` method in `google_search.rb` is not private, even though it's not meant to be used directly (it wasn't present in the old JSON annotations file).

This error could easily be repeated, so I thought explicitly defining the methods would make it difficult to introduce errors and provide a clear overview of available tools. What do you think?

Comment (andreibondarev) [reply lib/langchain/tool/base.rb:111]: @dghirardo In the case of `google_search.rb` specifically, I'm not even convinced that we need the `def self.execute_search` method?

**Comment by andreibondarev** [comment]
@dghirardo Btw -- @palladius told me that you guys met at the RubyDay! 🫶

File: lib/langchain/tool/base.rb (hunk 1 above)
Comment (dghirardo) [reply lib/langchain/tool/base.rb:111]: Yes, after looking at the code, it seems we don't need it. So, should we use `public_instance_methods` or keep the `FUNCTIONS` array?

**Comment by dghirardo** [comment]
@andreibondarev I am currently working on a pull request for the `tool_tailor` gem to add support for the `enum` parameter. What do you think about the [proposed solution](https://github.com/kieranklaassen/tool_tailor/issues/1)?

**Comment by dghirardo** [comment]
Hi @andreibondarev, `tool_tailor` has merged my pull request, so now the `enum` property is supported, updating the gem. However, I noticed another issue. When you have a parameter that is a collection (like an array or a hash), you can't describe its internal structure.

**Example** (Tavily tool):

```ruby
# @param exclude_domains [Array<String>] A list of domains to specifically exclude from the search results. Default is None, which doesn't exclude any domains.
```

While `[Array]` is supported, `[Array<String>]` is not.
So, it’s not possible to:
- Define that the array items must be Strings.
- Describe extra properties for the String values, such as `enum`.

You can only use the description field.

Even if we find a solution, I think nested parameters with multiple levels are not easy to describe with YARD comments.

**Comment by andreibondarev** [comment]
> Hi @andreibondarev, `tool_tailor` has merged my pull request, so now the `enum` property is supported, updating the gem. However, I noticed another issue. When you have a parameter that is a collection (like an array or a hash), you can't describe its internal structure.
> 
> **Example** (Tavily tool):
> 
> ```ruby
> # @param exclude_domains [Array<String>] A list of domains to specifically exclude from the search results. Default is None, which doesn't exclude any domains.
> ```
> 
> While `[Array]` is supported, `[Array<String>]` is not. So, it’s not possible to:
> 
> * Define that the array items must be Strings.
> * Describe extra properties for the String values, such as `enum`.
> 
> You can only use the description field.
> 
> Even if we find a solution, I think nested parameters with multiple levels are not easy to describe with YARD comments.

Hmm... I think we may want to consider some sort of a decorator pattern, a la:
```ruby
llm_callable :get_everything,
  description: "News Retriever: Search through millions of articles from over 150,000 large and small news sources and blogs",
  properties: {
    q: {
      type: :string,
      description: "Keywords or phrases to search for in the article title and body. Surround phrases with quotes (\") for exact match. Alternatively you can use the AND / OR / NOT keywords, and optionally group these with parenthesis. Must be URL-encoded."
  }

def get_everything(q:, ...)
  ...
```

**Comment by kieranklaassen** [comment]
Happy to add support for it in `tool_tailor`. Let me know if that would work. I think I want to support more use cases going forward as well.

**Comment by dghirardo** [comment]
Hi @andreibondarev, as agreed, since I have opened a new pull request with the new approach, I'm closing this one.

**Pull request closed without merging**